lib_eio_windows/sched.ml

```diff
@@ -0,0 +1,364 @@
+      Suspended.continue k ()                   (* A sleeping task is now due *)
+    | `Wait_until _ | `Nothing as next_due ->
+      let timeout =
+        match next_due with
+        | `Wait_until time ->
+          let time = Mtime.to_uint64_ns time in
+          let now = Mtime.to_uint64_ns now in
+          let diff_ns = Int64.sub time now in
+          (* Convert to seconds for Unix.select *)
+          let diff = Int64.(to_float diff_ns) /. 1_000_000_000. in
+          diff
+        | `Nothing -> (-1.)
+      in
+      if timeout < 0. && t.active_ops = 0 then (
+        (* Nothing further can happen at this point. *)
+        Lf_queue.close t.run_q;      (* Just to catch bugs if something tries to enqueue later *)
+        `Exit_scheduler
+      ) else (
+        Atomic.set t.need_wakeup true;
+        if Lf_queue.is_empty t.run_q then (
+          (* At this point we're not going to check [run_q] again before sleeping.
+             If [need_wakeup] is still [true], this is fine because we don't promise to do that.
+             If [need_wakeup = false], a wake-up event will arrive and wake us up soon. *)
+          Ctf.(note_hiatus Wait_for_work);
+          let cons _ fd acc = fd :: acc in
+          let read = FdMap.fold cons t.poll.to_read [] in
+          let write = FdMap.fold cons t.poll.to_write [] in
+          match Unix.select read write [] timeout with 
+          | exception Unix.(Unix_error (EINTR, _, _)) -> next t
+          | readable, writeable, _ ->
+            Ctf.note_resume system_thread;
+            Atomic.set t.need_wakeup false;
+            Lf_queue.push t.run_q IO;                   (* Re-inject IO job in the run queue *)
+            List.iter (ready t [ `W ]) writeable; 
+            List.iter (ready t [ `R ]) readable;
+            next t
+        ) else (
+          (* Someone added a new job while we were setting [need_wakeup] to [true].
+             They might or might not have seen that, so we can't be sure they'll send an event. *)
+          Atomic.set t.need_wakeup false;
+          Lf_queue.push t.run_q IO;                   (* Re-inject IO job in the run queue *)
+          next t
+        )
+      )
+
+let with_sched fn =
+  let run_q = Lf_queue.create () in
+  Lf_queue.push run_q IO;
+  let sleep_q = Zzz.create () in
+  (* Pipe's on Windows cannot be nonblocking through the OCaml API. *)
+  let eventfd_r, eventfd_w = Unix.socketpair ~cloexec:true Unix.PF_UNIX Unix.SOCK_STREAM 0 in
+  Unix.set_nonblock eventfd_r;
+  Unix.set_nonblock eventfd_w;
+  let eventfd = Rcfd.make eventfd_w in
+  let cleanup () =
+    Unix.close eventfd_r;
+    let was_open = Rcfd.close eventfd in
+    assert was_open
+  in
+  let poll = { to_read = FdMap.empty; to_write = FdMap.empty } in
+  let fd_map = Hashtbl.create 10 in
+  let t = { run_q; poll; poll_maxi = (-1); fd_map; eventfd; eventfd_r;
+            active_ops = 0; need_wakeup = Atomic.make false; sleep_q } in
+  let eventfd_ri : int = Obj.magic eventfd_r in
```

Comment:
I don't think file descriptors are ints on Windows.

Reply:
definitely not an integer; https://github.com/ocaml/ocaml/blob/trunk/otherlibs/unix/unixsupport.h#L43

Reply:
Yep, that's a little embarrassing ^^" Will fix soon

Reply:
The incorrect use of `Obj.magic` has been removed in https://github.com/ocaml-multicore/eio/pull/497/commits/3ab48eaf0e30a04f243aa991977c1b96ac55afb9 -- I'm now relying on compare `Unix.file_descr` for the set of thing to check to read and write.